Kickstarter-iOS/ViewModels/RootViewModel.swift

```diff
@@ -244,9 +253,65 @@ internal final class RootViewModel: RootViewModelType, RootViewModelInputs, Root
       .skipNil()
       .map { $0 as RootViewControllerIndex }
 
-    self.setBadgeValueAtIndex = activityViewControllerIndex
-      .takeWhen(self.viewDidLoadProperty.signal)
-      .map { index in (activitiesBadgeValue(), index) }
+    let lifecycleEvents = Signal.merge(
+      self.viewDidLoadProperty.signal,
+      self.applicationWillEnterForegroundSignal
+    )
+
+    let updateBadgeValueOnLifecycleEvents = activityViewControllerIndex
+      .takeWhen(lifecycleEvents)
+      .map { index in
+        (activitiesBadgeValue(with: AppEnvironment.current.application.applicationIconBadgeNumber), index)
+      }
+
+    let selectedIndexAndActivityViewControllerIndex = Signal.combineLatest(
+      self.selectedIndex,
+      activityViewControllerIndex
+    )
+
+    let badgeValueOnUserUpdated = self.currentUserUpdatedProperty.signal
+      .map { _ in AppEnvironment.current.currentUser?.unseenActivityCount }
+
+    let updateBadgeValueFromNotification = selectedIndexAndActivityViewControllerIndex
+      .takePairWhen(self.didReceiveBadgeValueSignal)
+
+    let updateBadgeValueOnUserUpdated = selectedIndexAndActivityViewControllerIndex
+      .takePairWhen(badgeValueOnUserUpdated)
+
+    let updateBadgeValueOnUserUpdatedOrFromNotification = Signal.merge(
+      updateBadgeValueOnUserUpdated,
+      updateBadgeValueFromNotification
+    )
+    .map(unpack)
+    .map { _, index, value in
+      (activitiesBadgeValue(with: value), index)
+    }
+
+    let clearBadgeValueOnUserSessionEnded = activityViewControllerIndex
+      .takePairWhen(self.userSessionEndedProperty.signal)
+      .map { index, _ in (activitiesBadgeValue(with: nil), index) }
+
+    let currentBadgeValue = MutableProperty<String?>(nil)
+
+    let clearBadgeValueOnActivitiesTabSelected = selectedIndexAndActivityViewControllerIndex.filter(==)
+      .flatMap { _, index in currentBadgeValue.producer.map { ($0, index) }.take(first: 1) }
+      .filter { value, _ in value != nil }
+      .map { _, index -> RootTabBarItemBadgeValueData in (nil, index) }
+
+    self.setBadgeValueAtIndex = Signal.merge(
+      updateBadgeValueOnLifecycleEvents,
+      updateBadgeValueOnUserUpdatedOrFromNotification,
+      clearBadgeValueOnUserSessionEnded,
+      clearBadgeValueOnActivitiesTabSelected
+    )
+
+    currentBadgeValue <~ self.setBadgeValueAtIndex.map { $0.0 }
+
+    self.updateUserInEnvironment = clearBadgeValueOnActivitiesTabSelected
```

Comment:
Should we add a check here for `AppEnvironment.current.currentUser != nil` before firing off the request?

Reply:
Hmm maybe but this does happen in the function that's called in the `switchMap` - `updatedUserWithClearedActivityCountProducer()` 🤷‍♂